### PR TITLE
bpo-6761: Fix __call__ documentation

### PR DESCRIPTION
**Pull request by andresdelfino**
https://bugs.python.org/issue6761

### CODE REVIEW

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

**Comment by andresdelfino** [comment]
@bitdancer could you take a look at this?

**Review by vstinner — CHANGES_REQUESTED** [review]
I prefer (arg1, arg2, ...) syntax rather than (arguments) syntax: IMHO (arg1, arg2, ...) is more explicit.

**Comment by andresdelfino** [comment]
I have made the requested changes; please review again

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
@csabella @JulienPalard: Would you mind to double check and/or merge this PR?

**Comment by vstinner** [comment]
> "needs backport 2.7"

I'm not super excited by continuing to enhance the 2.7 doc. IMHO it's good as it is. I mean, we are close to Python 2 end of life, and I prefer to read Python 3 doc to write Python 2 code. Well, it's just a comment, I'm also fine with backport to 2.7 if you want (but I expect a conflict).

**Comment by csabella** [comment]
This patch isn't incorrect, however, I'm wondering if something else is needed.

Looking at the original bug ticket, @bitdancer makes the point:

> FYI, all special methods are (now) looked up on the type for new style classes.

Taking this a step further, I tried the original example from the tracker with ``__str__`` instead of ``__call__`` and it produced the same result (of proving David's point about special methods being looked up on ``type(x)`` instead of ``x``.

This is already defined in the intro of this entire [3.3 section](https://docs.python.org/3.8/reference/datamodel.html#special-method-names):

> For instance, if a class defines a method named ``__getitem__()``, and x is an instance of this class, then x[i] is roughly equivalent to ``type(x).__getitem__(x, i)``.

I'm wondering if the `object.__call__` section should define what it does without referencing the ``x``.  The other sections (like ``__str__``) avoid using an example instance.

**Comment by andresdelfino** [comment]
> This patch isn't incorrect, however, I'm wondering if something else is needed.
> 
> Looking at the original bug ticket, @bitdancer makes the point:
> 
> > FYI, all special methods are (now) looked up on the type for new style classes.
> 
> Taking this a step further, I tried the original example from the tracker with `__str__` instead of `__call__` and it produced the same result (of proving David's point about special methods being looked up on `type(x)` instead of `x`.
> 
> This is already defined in the intro of this entire [3.3 section](https://docs.python.org/3.8/reference/datamodel.html#special-method-names):
> 
> > For instance, if a class defines a method named `__getitem__()`, and x is an instance of this class, then x[i] is roughly equivalent to `type(x).__getitem__(x, i)`.
> 
> I'm wondering if the `object.__call__` section should define what it does without referencing the `x`. The other sections (like `__str__`) avoid using an example instance.

I think you are right. Could you take a look at the new wording?

**Comment by csabella** [comment]
I'm definitely not the one who would be able to have the final say on the wording for this, but I wonder if something like this would work:

```
Called by ``object(arg1, arg2, ...)`` or an equivalent (such as a decorator) to invoke the instance as a function call.
```

**Comment by andresdelfino** [comment]
@vstinner could you take a look on the new wording I wrote based on Cheryl's comments on https://github.com/python/cpython/pull/7987#issuecomment-467026457 and also Cheryl's proposed wording on https://github.com/python/cpython/pull/7987#issuecomment-470110467 and tell us what you think?

**Comment by vstinner** [comment]
For me, obj(arg1, arg2, ...) is equivalant to (obj.__class__).__call__(obj, arg1, arg2, ...).

Usually, it's also equivalent to type(obj).__call__(arg1, arg2, ...) ... but I would prefer to use type() in the doc, since technically the builtin function can be overriden or a global variable with the same name can be defined.

File: Doc/reference/datamodel.rst
Comment (mblahay) [inline Doc/reference/datamodel.rst]: I know what is supposed to be conveyed here, and I find this more confusing than the original. Diving into the internals of how calls to instance methods work is detracting from the discussion of what __call__ does.

**Comment by andresdelfino** [comment]
@bitdancer Could you check the proposed wording out and commnet? :)

File: Doc/reference/datamodel.rst
Comment (andresdelfino) [reply Doc/reference/datamodel.rst]: Thanks for your insight! Before doing any changes, I'll wait for @bitdancer to give his input.

Comment (JulienPalard) [reply Doc/reference/datamodel.rst]: This current version don't work for me, "if this method is defined, the class receives the instance and all arguments of the call." looks wrong, It's not the class that receives the instance and all arguments.

I'd propose something like:

```Called when the instance is "called" as a function. Receives the instance and the function parameters.```

If you want to almost keep the example you could add something like:

```So ``obj(arg1, arg2, ...)`` calls ``__call__(arg1, arg2, ...)`` on ``obj``.```

Comment (merwok) [reply Doc/reference/datamodel.rst]: I don’t think that last suggestion (from Julien) works (call on obj?).

`type(x).__call__(x, arg1, ...)` is closer to what happens, and the call matches the way the method is defined (with a self parameter).

Comment (andresdelfino) [reply Doc/reference/datamodel.rst]: @merwok I agree. Made a change based on your suggestion, what do you think?

Comment (merwok) [reply Doc/reference/datamodel.rst]: @vstinner You said «I would prefer to [not] use type() in the doc, since technically the builtin function can be overriden» → that’s what we say «roughly»!  Is the latest version good for you or you would prefer `obj.__class__.__call__(obj, etc)`?

@csabella I’m not sure I understand your review; could you comment on the latest version?  Thanks!

**Review by vstinner — APPROVED** [review]
LGTM.

**Review by merwok — APPROVED** [review]

**Comment by merwok** [comment]
I updated the backports labels to get 3.9 and 3.8, and I think this is ready to merge unless another reviewer weighs in.

**Comment by vstinner** [comment]
Thanks @andresdelfino! It's a nice documentation enhancement. I'm fine with "roughly" ;-)

**Pull request merged**
